### PR TITLE
refactor: change default model from Merlinite to Granite

### LINKED ISSUE #2238: Switch default model from Merlinite to Granite
InstructLab has moved pretty heavily towards being a Granite-first project, yet many of the config defaults for models remain to be Merlinite. Opening this issue to start the conversation of switching over to Granite quanitized GGUFs and Safetensors, where applicable.

Merlinite can still be used in things like CI.

If folks agree I can convert this issue from a `question` to a `refactor` and implement it.

### PR DESCRIPTION
**Pull request by nathan-weinberg**
**Issue resolved by this Pull Request:**
Resolves #2238

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

**Comment by cdoern** [comment]
the functional failure seems to be that SDG is expecting merlinite still, could be that the flag default is off of the value in the cfg yaml is off.

**Review by jaideepr97 — COMMENTED** [review]
@nathan-weinberg we should probably also switch merlinite for granite here:
https://github.com/instructlab/instructlab/blob/main/src/instructlab/defaults.py#L14 and 
https://github.com/instructlab/instructlab/blob/main/src/instructlab/defaults.py#L17

so granite becomes an official "family" that we recognize and merlinite gets mapped to it instead of how it is right now

**Comment by nathan-weinberg** [comment]
> @nathan-weinberg we should probably also switch merlinite for granite here: https://github.com/instructlab/instructlab/blob/main/src/instructlab/defaults.py#L14 and https://github.com/instructlab/instructlab/blob/main/src/instructlab/defaults.py#L17
> 
> so granite becomes an official "family" that we recognize and merlinite gets mapped to it instead of how it is right now

Good catch! I will update the patchset to do this!

File: TROUBLESHOOTING.md
Comment (cdoern) [inline TROUBLESHOOTING.md:75]: one question I had: is it ok to switch the teacher model to granite? The idea behind the `simple` teacher being merlinite is that it is "smarter" than granite, part of me worries this will degrade the simple pipeline results further

File: src/instructlab/configuration.py
Comment (cdoern) [inline src/instructlab/configuration.py:157]: nice change, i don't know why mixtral wasn't here

File: src/instructlab/defaults.py
Comment (cdoern) [inline src/instructlab/defaults.py:18]: can you add `mistral: mixtral` in here? we have an identical copy of this map in SDG I didn't realize there was one here too

**Review by cdoern — COMMENTED** [review]
the only question i have is if we want the teacher model for the simple pipeline to be granite. We should probably do a comparison run to make sure it doesn't degrade results.

File: TROUBLESHOOTING.md
Comment (nathan-weinberg) [reply TROUBLESHOOTING.md:75]: I think we can leave the default teacher model to be `merlinite` - my only concern with it is the default model downloads - would we then be downloading three default models?
1. Merlinite GGUF as the teacher
2. Granite GGUF as the student
3. Mistral GGUF for SDG
If we're fine with this, would appreciate some pointers as to where exactly I need to make the changes to implement what I've laid out here

File: src/instructlab/defaults.py
Comment (nathan-weinberg) [reply src/instructlab/defaults.py:18]: Yessir will do!

File: TROUBLESHOOTING.md
Comment (RobotSail) [reply TROUBLESHOOTING.md:75]: We would definitely need to download at least 2 models - the teacher and the student, which need to be different for the sake of the student being able to catch up to the teacher. What you could do is make the mistral GGUF optional, and maybe down the road provide a way for the user to select which one they'd like.

Comment (nathan-weinberg) [reply TROUBLESHOOTING.md:75]: I think currently we are using Merlinite by default as both the student and the teacher and Mistral for SDG? I could be wrong, Charlie and @jaideepr97 would know for sure

Irregardless I'm happy to change this up a bit, but I don't want to stray too far from the goal of this PR

**Review by danmcp — COMMENTED** [review]
Changes LGTM.  I also pulled down the changes and checked for anything that might have been missed and didn't spot anything necessary.

Might be good to run the large CI job against it?

File: TROUBLESHOOTING.md
Comment (RobotSail) [reply TROUBLESHOOTING.md:75]: > student and the teacher and Mistral for SDG?

The teacher would be the model being used for SDG, so if it's teaching itself then that is a big problem.

Comment (nathan-weinberg) [reply TROUBLESHOOTING.md:75]: Oh I may be mistaken then, the current teacher might be Mistral? But in that case I'm not sure what @cdoern is referring to - am I accidentally switching it from Mistral to Granite? Because that was definitely not my intention

Comment (RobotSail) [reply TROUBLESHOOTING.md:75]: @nathan-weinberg You are correct, Merlinite is the version of Mistral-7B trained by IBM research. Granite is a separate foundation model from IBM. 

What we need to do is have the teacher model be more capable, which in the case of the simple pipeline has generally been the case. If we're trimming down on models that are downloaded, then we need to be sure that we still hold this property. I'm unfamiliar with the simple pipeline, but I don't believe we would be using base Mistral as a teacher model for SDG, so that may be the one to remove (we need to confirm this). 

But this pipeline is overall fairly old now, and IBM has Granite-3.0 8B and 3B, which could be potential replacements for either teacher & student models. We would just need to run some experiments to better understand the performance difference.

File: src/instructlab/defaults.py
Comment (jaideepr97) [inline src/instructlab/defaults.py:14]: I think the family should be `mistral` instead of `mixtral`, since that is the name of the company that makes the mixtral model and the mistral model

thoughts? @nathan-weinberg

Comment (nathan-weinberg) [reply src/instructlab/defaults.py:14]: I'm amenable to that for sure - @cdoern wdyt? since this was your original suggestion

File: TROUBLESHOOTING.md
Comment (cdoern) [reply TROUBLESHOOTING.md:75]: ok so there seems to be a bit of confusion here. Here is the current state of SDG upstream:

pipeline: full, model: mistral 7b instruct GGUf

pipeline: simple, model: merlinite 7b lab GGUF

My concern here is that in the docs and the defaults it seems that merl is being replaced w/ granite in SDG. I think we should:

a. download granite gguf as the default chat model
b. download merl gguf as the simple teacher model
c. download mistral gguf as the full teacher model

for _now_ we can download the all. However, I think we should conditionally download merlinite or mistral depending on the pipeline in the config eventually.

File: src/instructlab/defaults.py
Comment (cdoern) [reply src/instructlab/defaults.py:14]: mixtral is right, its what is used in SDG as the "family"

we map `mistral: mixtral` so that both types of models use the same templates.

Comment (jaideepr97) [reply src/instructlab/defaults.py:14]: Just because we use it elsewhere doesn't mean it's "right" though

Comment (cdoern) [reply src/instructlab/defaults.py:14]: sure, but I believe changing this would breakages in other places given that they depend on the "mixtral" model family

File: TROUBLESHOOTING.md
Comment (RobotSail) [reply TROUBLESHOOTING.md:75]: @cdoern Why is mistral GGUF being used as the full teacher model for the full pipeline while Merlinite is the full teacher model for the simple pipeline? Merlinite is produced as a result of fine-tuning Mistral, so there should ideally only be one model being used for this.

File: src/instructlab/defaults.py
Comment (nathan-weinberg) [reply src/instructlab/defaults.py:14]: I would prefer if we could discuss/do this in a separate PR - I am going to leave it as is to reduce the risk of causing a regression - happy to file a followup issue

File: TROUBLESHOOTING.md
Comment (RobotSail) [reply TROUBLESHOOTING.md:75]: Regardless of which models we're using, it seems like an easy solution here would be to have the following flow:

1. At `ilab init` the user would be prompted if they want to download all of the models now, or defer it until they need it

2. Then when a defined model doesn't exist on disk, or if they go to run SDG and the teacher models aren't present, they are prompted with a message like: "Mistral-7B-GGUF is required for this pipeline but appears to be missing. Would you like to download it?" And if they hit no, the pipeline is cancelled, else it's downloaded.

Comment (cdoern) [reply TROUBLESHOOTING.md:75]: @RobotSail Merlinite does not work with the full pipeline, consistently yields 0 generated samples and/or times out on the generation

Comment (cdoern) [reply TROUBLESHOOTING.md:75]: This PR is simply switching the default model to granite. 

I don't think we should make code changes outside of maybe downloading a third model, can we re-visit the download syntax in another issue/PR

Comment (nathan-weinberg) [reply TROUBLESHOOTING.md:75]: I'm going to go with this for now, @RobotSail i do wanna keep this convo going somewhere, either in Slack or a new issue, do you have a preference?

**Review by danmcp — APPROVED** [review]
LGTM, I'll suggest waiting for a couple days or 4+ approvals before letting it merge.

File: TROUBLESHOOTING.md
Comment (RobotSail) [reply TROUBLESHOOTING.md]: Yes, we'll make an issue/epic to describe the overall procedure.

**Review by cdoern — APPROVED** [review]

**Review by RobotSail — APPROVED** [review]
LGTM

**Review by kelbrown20 — APPROVED** [review]

**Pull request merged**
